src/metabase/data_app/api/data_app.clj

```diff
@@ -0,0 +1,108 @@
+  (cond-> {}
+    (some? status) (sql.helpers/where [:= :status (keyword status)])
+    (some? limit) (sql.helpers/limit limit)
+    (some? offset) (sql.helpers/offset offset)))
+
+(defn- filter-clauses-without-paging
+  "Given a where clause, return a clause that can be used to count."
+  [clauses]
+  (dissoc clauses :order-by :limit :offset))
+
+(api.macros/defendpoint :get "/"
+  "List all data apps. Optionally filter by status.
+
+  Takes `limit`, `offset` for pagination."
+  [_route-params
+   {:keys [status]} :- [:map
+                        [:status {:optional true} [:maybe [:enum "private" "published" "archived"]]]]]
+  (let [limit   (request/limit)
+        offset  (request/offset)
+        clauses (data-app-clauses status limit offset)]
+    {:data (t2/select :model/DataApp
+                      (sql.helpers/order-by clauses
+                                            [:created_at :desc]
+                                            [:id :asc]))
+     :total (t2/count :model/DataApp (filter-clauses-without-paging clauses))
+     :limit limit
+     :offset offset}))
+
+(api.macros/defendpoint :get "/:id"
+  "Get a specific data app by ID with latest definitions and release info."
+  [{:keys [id]} :- [:map [:id ms/PositiveInt]]]
+  (get-data-app id))
+
+(api.macros/defendpoint :post "/"
+  "Create a new data app with optional initial definition."
+  [_route _query body :- [:map
+                          [:name ms/NonBlankString]
+                          [:description {:optional true} [:maybe :string]]
+                          [:definition {:optional true} [:maybe :map]]]]
+  (api/create-check :model/DataApp body)
+  (data-app.models/create-app! (assoc body :creator_id api/*current-user-id*)))
+
+(api.macros/defendpoint :put "/:id"
+  "Update an existing data app."
+  [{:keys [id]} :- [:map [:id ms/PositiveInt]]
+   _query
+   body :- [:map
+            [:name {:optional true} ms/NonBlankString]
+            [:description {:optional true} [:maybe :string]]]]
+  (let [existing-data-app (get-data-app id)]
+    (api/update-check existing-data-app body)
+    (t2/update! :model/DataApp id body)
+    (get-data-app id)))
+
+(api.macros/defendpoint :delete "/:id"
+  "Soft delete a data app."
+  [{:keys [id]} :- [:map [:id ms/PositiveInt]]]
+  (get-data-app id)
+  (t2/update! :model/DataApp id {:status :archived})
+  api/generic-204-no-content)
+
+(api.macros/defendpoint :post "/:id/definition"
+  "Create a new definition version for a data app."
+  [{:keys [id]} :- [:map [:id ms/PositiveInt]]
+   _query
+   body]
+  (let [data-app (get-data-app id)]
+    (api/update-check data-app {})
+    (data-app.models/new-definition! id body)))
+
+(api.macros/defendpoint :post "/:id/release"
+  "Release a specific definition version of a data app."
+  [{:keys [id]} :- [:map [:id ms/PositiveInt]]
+   _query
+   body :- [:map
+            [:definition_id ms/PositiveInt]]]
+  ;; should we assume that we always release the latest definition?
```

Comment:
Yeah I think that's fine for now. I'm thinking that the FE shouldn't even need to know about `definition_id` for now, it's a lower level implementation detail.